libsolidity/codegen/ir/IRGeneratorForStatements.cpp

```diff
@@ -1719,3 +1701,191 @@ Type const& IRGeneratorForStatements::type(Expression const& _expression)
 	solAssert(_expression.annotation().type, "Type of expression not set.");
 	return *_expression.annotation().type;
 }
+
+bool IRGeneratorForStatements::visit(TryStatement const& _tryStatement)
+{
+	solAssert(!m_returnInfo.has_value(), "");
+
+	_tryStatement.externalCall().accept(*this);
+	solAssert(m_returnInfo.has_value(), "");
+	solAssert(m_returnInfo.value().functionCall == &_tryStatement.externalCall(), "");
+
+	ReturnInfo const returnInfo = move(*m_returnInfo);
+	m_returnInfo.reset();
+
+	string const trySuccessCondition = m_context.trySuccessConditionVariable(_tryStatement.externalCall());
+
+	m_code << "switch iszero(" << trySuccessCondition << ")\n";
+
+	m_code << "case 0 { // success case\n";
+	TryCatchClause const& successClause = *_tryStatement.clauses().front();
+	decodeTryCallReturnParameters(successClause, returnInfo);
+	successClause.block().accept(*this);
+	m_code << "}\n";
+
+	m_code << "default { // failure case\n";
+	handleCatch(_tryStatement);
+	m_code << "}\n";
+
+	return false;
+}
+
+string IRGeneratorForStatements::decodeReturnParameters(ReturnInfo const& _returnInfo)
```

Comment:
I'm wondering if the code would be better readable if we did not have this function. I think it's always dangerous if we inject code that is generated somewhere else and declares yul variables in "another" scope.

Reply:
We are calling this function twice. Once in `IRGeneratorForStatements::decodeTryCallReturnParameters` where it would make sense to not return the code but to let this function actually append the code to `m_code`.

But the second place is the nested template, and in order to avoid code duplication, I *think* we should preserve this function as it is.

Reply:
I just think the required code duplication would be the lesser evil than injecting code that "magically" declares a new local variable.

Reply:
If you do not know the context, the fact that the template above is just using
```
<decodeReturnParameters>
```
does not make you assume that this actually declares a variable.

Reply:
done.